src/main/flight/kalman.c

```diff
@@ -0,0 +1,159 @@
+{
+    setPoint[axis] = rate;
+}
+
+void gyroKalmanInitialize(void)
+{
+    memset(&varStruct, 0, sizeof(varStruct));
+    gyroKalmanInitAxis(&kalmanFilterStateRate[X], gyroConfig()->kalman_q);
+    gyroKalmanInitAxis(&kalmanFilterStateRate[Y], gyroConfig()->kalman_q);
+    gyroKalmanInitAxis(&kalmanFilterStateRate[Z], gyroConfig()->kalman_q);
+
+    varStruct.w = gyroConfig()->kalman_w * 8;
+    varStruct.inverseN = 1.0f / (float)(varStruct.w);
+}
+
+static void update_kalman_covariance(float *gyroRateData)
+{
+    varStruct.xWindow[varStruct.windex] = gyroRateData[X];
+    varStruct.yWindow[varStruct.windex] = gyroRateData[Y];
+    varStruct.zWindow[varStruct.windex] = gyroRateData[Z];
+
+    varStruct.xSumMean += varStruct.xWindow[varStruct.windex];
+    varStruct.ySumMean += varStruct.yWindow[varStruct.windex];
+    varStruct.zSumMean += varStruct.zWindow[varStruct.windex];
+    varStruct.xSumVar = varStruct.xSumVar + (varStruct.xWindow[varStruct.windex] * varStruct.xWindow[varStruct.windex]);
+    varStruct.ySumVar = varStruct.ySumVar + (varStruct.yWindow[varStruct.windex] * varStruct.yWindow[varStruct.windex]);
+    varStruct.zSumVar = varStruct.zSumVar + (varStruct.zWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.xySumCoVar = varStruct.xySumCoVar + (varStruct.xWindow[varStruct.windex] * varStruct.yWindow[varStruct.windex]);
+    varStruct.xzSumCoVar = varStruct.xzSumCoVar + (varStruct.xWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.yzSumCoVar = varStruct.yzSumCoVar + (varStruct.yWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.windex++;
+    if (varStruct.windex >= varStruct.w)
+    {
+        varStruct.windex = 0;
+    }
+    varStruct.xSumMean -= varStruct.xWindow[varStruct.windex];
+    varStruct.ySumMean -= varStruct.yWindow[varStruct.windex];
+    varStruct.zSumMean -= varStruct.zWindow[varStruct.windex];
+    varStruct.xSumVar = varStruct.xSumVar - (varStruct.xWindow[varStruct.windex] * varStruct.xWindow[varStruct.windex]);
+    varStruct.ySumVar = varStruct.ySumVar - (varStruct.yWindow[varStruct.windex] * varStruct.yWindow[varStruct.windex]);
+    varStruct.zSumVar = varStruct.zSumVar - (varStruct.zWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.xySumCoVar = varStruct.xySumCoVar - (varStruct.xWindow[varStruct.windex] * varStruct.yWindow[varStruct.windex]);
+    varStruct.xzSumCoVar = varStruct.xzSumCoVar - (varStruct.xWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+    varStruct.yzSumCoVar = varStruct.yzSumCoVar - (varStruct.yWindow[varStruct.windex] * varStruct.zWindow[varStruct.windex]);
+
+    varStruct.xMean = varStruct.xSumMean * varStruct.inverseN;
+    varStruct.yMean = varStruct.ySumMean * varStruct.inverseN;
+    varStruct.zMean = varStruct.zSumMean * varStruct.inverseN;
+
+    varStruct.xVar = fabsf(varStruct.xSumVar * varStruct.inverseN - (varStruct.xMean * varStruct.xMean));
```

Comment:
This is not correct. 
Variance is `average((x - mean)^2)`
However this algorithm computes `average(x)^2 - average(mean)^2`